src/serac/numerics/functional/isotropic_tensor.hpp

```diff
@@ -0,0 +1,457 @@
+ *   d(antisym(A)_{ij}) / d(A_{kl})
+ * @tparam m the dimension
+ */
+template <int m>
+SERAC_HOST_DEVICE constexpr auto AntisymmetricIdentity()
+{
+  return isotropic_tensor<double, m, m, m, m>{0.0, 0.0, 1.0};
+}
+
+/**
+ * @brief scalar multiplication
+ *
+ * @tparam S the type of the left operand, `scale`
+ * @tparam T the type of the isotropic tensor
+ * @tparam m the dimension of each rank-4 isotropic tensor
+ * @param scale the value that multiplies each entry of I (from the left)
+ * @param I the isotropic tensor being scaled
+ * @return a new isotropic tensor equal to the product of scale * I
+ */
+template <typename S, typename T, int m>
+SERAC_HOST_DEVICE constexpr auto operator*(S scale, isotropic_tensor<T, m, m, m, m> I)
+{
+  return isotropic_tensor<decltype(S{} * T{}), m, m, m, m>{I.c1 * scale, I.c2 * scale, I.c3 * scale};
+}
+
+/// @overload
+template <typename S, typename T, int m>
+SERAC_HOST_DEVICE constexpr auto operator*(isotropic_tensor<S, m, m, m, m> I, T scale)
+{
+  return isotropic_tensor<decltype(S{} * T{}), m, m, m, m>{I.c1 * scale, I.c2 * scale, I.c3 * scale};
+}
+
+/**
+ * @brief addition of isotropic tensors
+ *
+ * @tparam S the type of the left isotropic tensor
+ * @tparam T the type of the right isotropic tensor
+ * @tparam m the dimension of each rank-4 isotropic tensor
+ * @param I1 the left operand
+ * @param I2 the right operand
+ * @return a new isotropic tensor equal to the sum of I1 and I2
+ */
+template <typename S, typename T, int m>
+SERAC_HOST_DEVICE constexpr auto operator+(isotropic_tensor<S, m, m, m, m> I1, isotropic_tensor<T, m, m, m, m> I2)
+{
+  return isotropic_tensor<decltype(S{} + T{}), m, m, m, m>{I1.c1 + I2.c1, I1.c2 + I2.c2, I1.c3 + I2.c3};
+}
+
+/**
+ * @brief difference of isotropic tensors
+ *
+ * @tparam S the type of the left isotropic tensor
+ * @tparam T the type of the right isotropic tensor
+ * @tparam m the dimension of each rank-4 isotropic tensor
+ * @param I1 the left operand
+ * @param I2 the right operand
+ * @return a new isotropic tensor equal to the difference of I1 and I2
+ */
+template <typename S, typename T, int m>
+SERAC_HOST_DEVICE constexpr auto operator-(isotropic_tensor<S, m, m, m, m> I1, isotropic_tensor<T, m, m, m, m> I2)
+{
+  return isotropic_tensor<decltype(S{} - T{}), m, m, m, m>{I1.c1 - I2.c1, I1.c2 - I2.c2, I1.c3 - I2.c3};
+}
+
+/**
+ * @brief double-dot product between an isotropic and (nonisotropic) tensor
+ *
+ * @tparam S the types stored in isotropic tensor
+ * @tparam T the types stored in tensor
+ * @tparam m the dimension of each extent of I
+ * @param I the left operand
+ * @param A the (full) right operand
+ * @return a new tensor equal to the index notation expression:
+ *    output(...) := I(i,j) * A(i,j,...)
+ */
+template <typename S, typename T, int m, int... n>
+SERAC_HOST_DEVICE constexpr auto ddot(const isotropic_tensor<S, m, m, m, m>& I, const tensor<T, m, m, n...>& A)
```

Comment:
Same comment as `sqnorm`. If this is common (e.g. a Matlab/Mathematica call), I'm fine with it. Otherwise, it should be more explicit (e.g. `double_dot`).

Reply:
hah, I was trying to remember where this came from, it's LaTeX for 2 dots over a character (not a useful name here).